### PR TITLE
Properly restore GrantedAuthorities

### LINKED ISSUE #41: Access via API Token won't use group information
Albeit a user is in an administrative group, that user is not able to use the REST API via API Token and gets a `403 Forbidden` response. The Web UI works as expected. Also the `/api/json` page when opening it up in the browser.

But when trying to use it with the API Token, the HTTP response body is telling that the `user is missing the Overall/Read permission`.

I reckon this might be connected to the fact that group information is ephemeral and just connected to the ID Token? This in turn renders API Tokens (and probably also SSH key based access) a bit useless, when user permissions are purely managed via groups.

If this is really the cause of the 403, would it maybe be feasible to add some caching to the ID Token group information? Maybe connected to the lifetime of the ID Token. That way, one could at least use the API Token for a certain amount of time.

```
$ http -a user:token https://jenkins/api/json
HTTP/1.1 403 Forbidden
Cache-Control: no-cache,no-store,must-revalidate
Content-Encoding: gzip
Content-Length: 1859
Content-Type: text/html;charset=utf-8
Date: Thu, 26 Jul 2018 10:08:24 GMT
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Server: Jetty(9.4.z-SNAPSHOT)
Set-Cookie: JSESSIONID.f9ff99b0=node0bhlg184ovzfu1vr9mglaapt5j37158.node0;Path=/;Secure;HttpOnly
X-Content-Type-Options: nosniff
X-Frame-Options: sameorigin
X-Hudson: 1.395
X-Hudson-CLI-Port: 50000
X-Hudson-Theme: default
X-Instance-Identity: MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAm7z6NAvVLlKLhOAs7tIckjHQyMXzWnoRx9fmfKeAB2aCZ61Xto7t91DTY4pMFMNDgFjVdTSKW3pmKGMflb/fgEnJRK/o2XO0IptQoDGWPn8pStZJlkEnweIN93k0epfXi8BL76V0RgkoshZmi8iC6ITm+Ta3N9AeUxSyzed9rwAoRrSecC2SwJa75DqxnXcauBGvFQYM71EOM2dVFtedDsxNr8qPVWofJxF/gcSZlUPj74HCySWd1b/WCiojDd23O0m/Y5MDX/8bT4r0hf0p4ytbo4erbEx1khYBWAHhnbBHyFpVeben0DEG2+3kjYWUX4SNyGh8EFeYvs54uElHEQIDAQAB
X-Jenkins: 2.121.2
X-Jenkins-CLI-Port: 50000
X-Jenkins-CLI2-Port: 50000
X-Jenkins-Session: c60d4fbc
X-Permission-Implied-By: hudson.security.Permission.GenericRead
X-Permission-Implied-By: hudson.model.Hudson.Administer
X-Required-Permission: hudson.model.Hudson.Read
X-You-Are-Authenticated-As: user
X-You-Are-In-Group-Disabled: JENKINS-39402: use -Dhudson.security.AccessDeniedException2.REPORT_GROUP_HEADERS=true or use /whoAmI to diagnose
```

### LINKED ISSUE #87: OIC user not able to make API calls
I am using Jenkins Login with Openid Connect (Amazon Cognito). I am able to use group-based authorization. I can see my groups in Granted Authorities: authenticated,<cognito group>

But when I try making API call, it gives me a "403" error saying "Missing overall read permissions". 

I am using API token created using <Jenkins url>/configure/me

It is only allowing in case I give "Read" access to the anonymous group in Jenkins, which I couldn't give in my production environment.

Issue::

Jenkins is not able to read neither SSO users authorized in groups nor in the authenticated group.
Can't we use SSO users to make API calls

### PR DESCRIPTION
**Pull request by jilyaluk**
I'm really eager to see #99 officially packaged, and original PR seems a bit not maintained, fixing review comments in this PR.

Hope that @twz123 don't mind.

@LinuxSuRen what did you mean in [this comment](https://github.com/jenkinsci/oic-auth-plugin/pull/99#discussion_r672906737)? What's wrong with lines 343 and 345 at the moment?

-- original description --

In the custom UserDetailsService implementation, getAllProperties() was
used to restore the GrantedAuthorities, which won't work for non-admin
users. Replace that with a call to getProperty(), which behaves as
expected.

Fixes #41.
Fixes #87.

### CODE REVIEW

**Comment by twz123** [comment]
@jilyaluk Sorry, I don't track Github notifications that well (but your mention did help). Didn't realize that I got a review after one year ;) If you wanna take over from here, please do! :+1:

**Comment by jilyaluk** [comment]
Glad to hear! Thanks for your original PR :)

**Comment by gammpamm** [comment]
We urgently need this fix as well. I tested the PR in Jenkins and it works.

**Comment by fbuchmeier-abi** [comment]
Great job @jilyaluk, this issue prevents us from adopting OIDC in Jenkins in production as well. Hope it gets merged soon!

We are happy to help as we can if there are any changes that need to be made.

**Comment by michaelbenoit** [comment]
We would need that fix. Is there any chance that it will be merged ?

**Comment by gregoryboue** [comment]
Hi, any news of this we are waiting for this PR for long time

**Comment by twz123** [comment]
I think that the only way to get this fix right now is to package the plugin on your own.

Incremental builds (#91) have also not been merged yet, and the plugin is [up for adoption](https://github.com/jenkinsci/oic-auth-plugin/pull/99#issuecomment-1200867148).

An easy way to do it with Docker:

```console
❯ docker build - <<EOF
FROM maven:3-jdk-8-slim as oic-auth
RUN mkdir -p /workspace/oic-auth
RUN curl -sSL https://api.github.com/repos/jilyaluk/oic-auth-plugin/tarball/fix-loadbyusername | tar xzv --strip-components=1 -C /workspace/oic-auth
RUN cd /workspace/oic-auth && mvn dependency:go-offline
RUN cd /workspace/oic-auth && mvn verify
EOF

[... snip ...]
Successfully built 2d33d17b1cd0

❯ docker run --rm --entrypoint tar 2d33d17b1cd0 -C workspace/oic-auth/target -cv oic-auth.hpi > oic-auth.hpi.tar
oic-auth.hpi

❯ tar tvf oic-auth.hpi.tar
-rw-r--r-- root/root    634204 2022-08-05 12:15 oic-auth.hpi
```

**Comment by twz123** [comment]
@jilyaluk mind to resolve the conflicts? Apparently there's a [new maintainer in the house](https://github.com/jenkins-infra/repository-permissions-updater/pull/2817). :tada:

**Comment by jilyaluk** [comment]
That's great news!

I resolved the conflicts. @Skiepp could you take a look?

**Comment by Skiepp** [comment]
@jilyaluk I just updated the plugin for being version compatible and restore properly the group permission
I'm not planning on performing updates and I do not have anymore permission

Feel free to adopt this plugin

**Comment by michael-doubez** [comment]
@jilyaluk Compilation fails

    [2022-10-18T18:59:58.904Z] [ERROR] /home/jenkins/workspace/Plugins_oic-auth-plugin_PR-115/src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:[386,85] incompatible types: org.springframework.security.core.GrantedAuthority[] cannot be converted to java.util.Collection<? extends org.springframework.security.core.GrantedAuthority>

I'll make a release soon before sorting the multiple PR about group parsing.

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 1]
```diff
@@ -381,17 +381,11 @@ public UserDetails loadUserByUsername(String username) throws UsernameNotFoundEx
...
-						for (UserProperty prop: props) {
-							LOGGER.fine("loadUserByUsername in createSecurityComponents called, prop of type: " + prop.getClass().toString());
-							if (prop instanceof OicUserProperty) {
-								OicUserProperty oicProp = (OicUserProperty) prop;
-								LOGGER.fine("loadUserByUsername in createSecurityComponents called, oic prop found with username: " + oicProp.getUserName());
-								auths = oicProp.getAuthoritiesAsGrantedAuthorities();
-								LOGGER.fine("loadUserByUsername in createSecurityComponents called, oic prop with auths size: " + auths.size());
-							}
+						OicUserProperty oicProp = u.getProperty(OicUserProperty.class);
+						GrantedAuthority[] auths = new GrantedAuthority[0];
```
Comment (michael-doubez) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:385]: Should be

    List<GrantedAuthority> auths = new ArrayList<>();

**Comment by michael-doubez** [comment]
I'll merge the PR and FIX the code.

**Comment by twz123** [comment]
Thanks everyone! Long standing issue resolved!

**Pull request merged**